# EARSM

Explicit algebraic Reynolds stress models (EARSM) are somewhere in between $k$-$\varepsilon$ and full RSM; the best of both world, i.e. fast and stable, but more accurate than your usual $k$-$\varepsilon$. 

The two main components are:

- The implicit algrebraic Reynolds stress equation (with LRR-QI model for pressure strain, which is most typically used):

$$
\left(c_1 - 1 + \frac{\mathcal{P}}{\varepsilon}\right) a = - \frac{8}{15} S + \frac{7 c_2 + 1}{11} \left( a \Omega - \Omega a \right) - \frac{5 - 9 c_2}{11} \left(a S + S a + 2/3 tr(aS) I \right)
$$

- The Pope universal anisotropy equation:

$$
a = \sum \beta_i T_i
$$

The set of tensors, $T_i$, are slightly different depending on the source:


$$
\begin{matrix}
 & Pope~(1975) & Apsley+Leschziner~(1998) & Wallin~(2000) \\
 & Taulbee~(1992) &  &  \\
 & Gatski+Speziale~(1993) &  &  \\
T_1 = & S  &  S  &  S\\
T_2 =   &  S \Omega - \Omega S &  \Omega S - S \Omega & S^2 - \frac{1}{3} II_S \mathbf{I}\\
T_3 =   &  S^2 - \frac{1}{3} II_S \mathbf{I}  &  S^2 - \frac{1}{3} II_S \mathbf{I} & \Omega^2 - \frac{1}{3} II_\Omega \mathbf{I} \\
T_4 = & \Omega^2 - \frac{1}{3} II_\Omega \mathbf{I}  &  \Omega^2 - \frac{1}{3} II_\Omega \mathbf{I}  & S \Omega - \Omega S \\
T_5 = & \Omega S^2 - S^2 \Omega   &  \Omega S^2 - S^2 \Omega  & S^2 \Omega - \Omega S^2 \\
T_6 = &  S \Omega^2 + \Omega^2 S - \frac{2}{3} IV \mathbf{I} & S \Omega^2 + \Omega^2 S - II_\Omega S - \frac{2}{3} IV \mathbf{I}   & S \Omega^2 + \Omega^2 S - \frac{2}{3} IV \mathbf{I} \\
T_7 = & \Omega S \Omega^2 - \Omega^2 S \Omega  &  \Omega S \Omega^2 - \Omega^2 S \Omega  & S^2 \Omega^2 + \Omega^2 S^2 - \frac{2}{3} V \mathbf{I} \\
T_8 = & S \Omega S^2 - S^2 \Omega S  &  S \Omega S^2 - S^2 \Omega S  & S \Omega S^2 - S^2 \Omega S \\
T_9 = & S^2 \Omega^2 + \Omega^2 S^2 - \frac{2}{3} V \mathbf{I}  &  S^2 \Omega^2 + \Omega^2 S^2 - \frac{2}{3} V \mathbf{I}  & \Omega S \Omega^2 - \Omega^2 S \Omega \\
T_{10} = & \Omega S^2 \Omega^2 - \Omega^2 S^2 \Omega  &  \Omega S^2 \Omega^2 - \Omega^2 S^2 \Omega  & \Omega S^2 \Omega^2 - \Omega^2 S^2 \Omega \\
\end{matrix}
$$


We shall use the ones of Wallin (2000), if not stated otherwise:






Note, that we use the following definitions:

$$
\begin{align}
S_{ij} &= \frac{1}{2} \frac{k}{\varepsilon} \left(\frac{\partial U_i}{\partial x_j} + \frac{\partial U_j}{\partial x_i} \right)  \\
\Omega_{ij} &= \frac{1}{2} \frac{k}{\varepsilon} \left(\frac{\partial U_i}{\partial x_j} - \frac{\partial U_j}{\partial x_i} \right)  \\
\end{align}
$$

The invariants are:

$$
II_S \equiv S_{ij} S_{ji} , \quad II_\Omega \equiv \Omega_{ij} \Omega_{ji} 
$$

$$
III \equiv S_{ij} S_{jk} S_{ki} , \quad IV \equiv S_{ij} \Omega_{jk} \Omega_{ki} , \quad V \equiv S_{ij} S_{jk} \Omega_{km} \Omega_{mi} 
$$


The shear parameter (Apsley+Leschziner 1998 and van der Laan 2014) is defined as:

$$
\sigma \equiv \frac{k}{\varepsilon} \sqrt{\frac{\partial U_i}{\partial x_j} \frac{\partial U_i}{\partial x_j}} =  \sqrt{(S_{ij}+\Omega_{ij}) (S_{ij}+\Omega_{ij})} = \sqrt{II_S - II_\Omega}
$$

- Note 1: The latter equality comes from the trace of symmetric times anti-symmetrix matrix is 0 (c.f. Pope appendix).
- Note 2: Pope (1975) and Wallin (2000) define $\sigma=\sqrt{II_S/2}$, while Taulbee (1992) defines $\sigma = \sqrt{II_S}$, but neither of these definitions shall be used.

## History of EARSM

All EARS models boils down to the description of the $\beta$ coefficients and many different models have been suggested in the history.

- Pope (1975): The $a = \sum_i \beta_i T_i$ is first formulated. Showed that for 3D there are 10 tensor groups and 5 invariants, while for 2D there are 3 tensor groups and 2 invariants. The 2D solution is derived (leaves $N$ implicit, i.e. calculate $N$ using values from timestep $n+1$).
- Taulbee (1992): The simplified (i.e. $c_2=5/9$) 3D solution is derived (leaves $N$ implicit).
- Gatski and Speziale (1993): The general 3D solution is derived (fixes $N=1$).
- Johansson & Wallin (1996): The self-consistent (i.e. does NOT leave $N$ implicit, but calculates with values from timestep $n$ using closed analytical cubic solution), simplified 2D solution is derived.
- Girimaji (1996): The self-consistent, general 2D solution is derived.

## EARSM in 2D

If there is no variation in one direction (e.g. flow over backward-facing step, flow over wavy surface, homogeneous ABL, etc.) and the velocity in that direction is 0, then the tensor group simplifies significantly:


$$
\begin{matrix}
 & Pope~(1975) & Apsley+Leschziner~(1998) & Wallin~(2000) \\
 & Taulbee~(1992) &  &  \\
 & Gatski+Speziale~(1993) &  &  \\
T_1 = & S  &  S  &  S\\
T_2 =   &  S \Omega - \Omega S 
        &  \Omega S - S \Omega 
        & S^2 - \frac{1}{3} II_S \mathbf{I}\\
T_3 =   &  S^2 - \frac{1}{3} II_S \mathbf{I}  
        &  S^2 - \frac{1}{3} II_S \mathbf{I} 
        & \frac{II_\Omega}{II_S} T_2 \\
T_4 = & \frac{II_\Omega}{II_S}T_3
      & \frac{II_\Omega}{II_S}T_3  
      & S \Omega - \Omega S \\
T_5 = & 0   
      & 0  
      & 0 \\
T_6 = & II_\Omega T_1 
      & 0   
      & II_\Omega T_1 \\
T_7 = & -\frac{1}{2} II_\Omega T_2  
      & \frac{1}{2} II_\Omega T_2  
      & II_\Omega T_2 \\
T_8 = & -\frac{1}{2} II_\Omega T_2  
      & \frac{1}{2} II_\Omega T_2  
      & -\frac{1}{2} II_\Omega T_4 \\
T_9 = & II_\Omega T_3
      & II_\Omega T_3  
      & - \frac{1}{2} II_\Omega T_4 \\
T_{10} = & 0  
         & 0  
         & 0 \\
\end{matrix}
$$


There are only three independent tensors in 2D (all others are either 0 or functions of the three). For example $T_5 = 0$ can be derived using the following expressions valid in 2D (Pope 1975, appendix):

$$
S^2 = \frac{1}{2}  II_S I_2 \quad , \quad \Omega^2 = \frac{1}{2}  II_\Omega I_2
$$

$$
S \Omega S = - \frac{1}{2} \Omega II_S \quad , \quad \Omega S \Omega = - \frac{1}{2} S II_\Omega
$$

Apsley groups:

$$
\begin{align}
T_5 &\equiv \Omega S^2 - S^2 \Omega \\
    &= \Omega \left(\frac{1}{2}  II_S I_2 \right) - \left(\frac{1}{2}  II_S I_2 \right) \Omega \\
    &= 0
\end{align}
$$


$$
\begin{align}
T_6 &\equiv S \Omega^2 + \Omega^2 S - II_\Omega S - \frac{2}{3} IV \mathbf{I} \\
    &= S \left(\frac{1}{2}  II_\Omega I_2 \right) + \left(\frac{1}{2}  II_\Omega I_2 \right) S - II_\Omega S - \frac{2}{3} IV \mathbf{I} \\
    &= 0
\end{align}
$$

$$
I_2 = \begin{bmatrix} 1 & 0 & 0 \\ 0 & 0 & 0 \\ 0 & 0 & 1 \end{bmatrix} \quad (\text{no variation in y-direction})
$$


The invariants are:

$$
II_S \equiv S_{ij} S_{ji} , \quad II_\Omega \equiv \Omega_{ij} \Omega_{ji} 
$$

$$
III = 0 , \quad IV = 0 , \quad V \neq 0 
$$

Wallin writes that $II_\Omega < 0$ always.

The strain, mean rotation and anisotropy tensors are (assuming no variation in $y$-direction, we are in the $x$-$z$ plane):

$$
S_{ij} = \begin{bmatrix} S_{11} & S_{12} & S_{13} \\ S_{12} & 0 & S_{23} \\ S_{13} & S_{23} & -S_{11} \end{bmatrix}  \quad , \quad
\Omega_{ij} = \begin{bmatrix} 0 & \Omega_{12} & \Omega_{13} \\ -\Omega_{12} & 0 & \Omega_{23} \\ -\Omega_{13} & -\Omega_{23} & 0 \end{bmatrix} 
 \quad , \quad
a_{ij} = \begin{bmatrix} a_{11} & a_{12} & a_{13} \\ a_{12} & a_{22} & a_{23} \\ a_{13} & a_{23} & -a_{11}-a_{22} \end{bmatrix} 
$$

- The tensors are not simplified significantly, so we might as well keep them in 3D formulation in the code. Can of course also use the 3D model for 1D and 2D flows, but the coding is just more complicated than necessary.
- A single-column model (SCM) is a subcase of 2D, where we also assume no variation in $x$-direction. This also means that $W = 0$, because of continuity, and the tensors will simplify.
- A subcase of SCM is "simple shear flow", where we also assume uni-directional flow, e.g. log-law or channel flow.

$$
(T_4)_{22} = (S_{12} \Omega_{12} - S_{23} \Omega_{23}) - (- S_{12} \Omega_{12} - S_{23} \Omega_{23}) = 2 S_{12} \Omega_{12}
$$

## EARSM in parallel flow (aka. unidirectional 1D flow)

If there is only variation in one direction (e.g. channel flow, neutral ASL, homegeneous shear flow), i.e. only $dU/dz \neq 0$.

$$
S_{ij} =  \begin{bmatrix} 0 & 0 & \frac{1}{2}\sigma \\ 0 & 0 & 0 \\ \frac{1}{2}\sigma & 0 & 0 \end{bmatrix} \quad , \quad \Omega_{ij} =  \begin{bmatrix} 0 & 0 & \frac{1}{2}\sigma \\ 0 & 0 & 0 \\ -\frac{1}{2}\sigma & 0 & 0 \end{bmatrix}
$$


- $\sigma \equiv \sqrt{II_S - II_\Omega} = \frac{k}{\varepsilon} \frac{dU}{dz}$.
- $II_S = -II_\Omega = \frac{1}{2} \sigma^2$
- $III = IV = 0$
- $V = -\frac{1}{8} \sigma^4$
- $P/\varepsilon \equiv -a_{ij} S_{ji} = - 2 a_{13} S_{13} = -a_{13} \sigma$, hence $a_{13}$ is the only anisotropy component contributing to $P/\varepsilon$ (Wallin 2000, p.96).
- $a_{33}$ is the only anisotropy component contributing to turbulent diffusion term (Wallin 2000, p.96)
- $S^2 = \frac{1}{2} II_S I_2$ and $\Omega^2 = \frac{1}{2} II_\Omega I_2$, hence $S^2 = - \Omega^2$.
- $ S \Omega S = - \frac{1}{2} \Omega II_S$ and  $\Omega S \Omega = - \frac{1}{2} S II_\Omega$.


$$
\begin{matrix}
 & Pope~(1975) & Apsley+Leschziner~(1998) & Wallin~(2000) \\
 & Taulbee~(1992) &  &  \\
 & Gatski+Speziale~(1993) &  &  \\
T_1 = & S  &  S  &  S\\
T_2 =   &  S \Omega - \Omega S 
        &  \Omega S - S \Omega 
        & S^2 - \frac{1}{3} II_S \mathbf{I}\\
T_3 =   &  S^2 - \frac{1}{3} II_S \mathbf{I}  
        &  S^2 - \frac{1}{3} II_S \mathbf{I} 
        & -T_2 \\
T_4 = & -T_3
      & -T_3  
      & S \Omega - \Omega S \\
T_5 = & 0   
      & 0  
      & 0 \\
T_6 = & II_\Omega T_1 
      & 0   
      & II_\Omega T_1 \\
T_7 = & -\frac{1}{2} II_\Omega T_2  
      & \frac{1}{2} II_\Omega T_2  
      & II_\Omega T_2 \\
T_8 = & -\frac{1}{2} II_\Omega T_2  
      & \frac{1}{2} II_\Omega T_2  
      & -\frac{1}{2} II_\Omega T_4 \\
T_9 = & II_\Omega T_3
      & II_\Omega T_3  
      & - \frac{1}{2} II_\Omega T_4 \\
T_{10} = & 0  
         & 0  
         & 0 \\
\end{matrix}
$$



## Self-conistent simplified 2D EARSM (Wallin 2000)

It is called "simplified" in Wallin (2000), because a clever choice of model constants makes the equations much simpler (more details in the derivation at the end of the notebook), while "self-consistent" means that we calculate $P/\varepsilon$ (or equivalently $N$) as a part of the solution instead of fixing it or leaving it implicit. In summary the model is applied as:

- From last iteration, we have $U$, $V$, $k$ and $\varepsilon$ $\rightarrow$ we can get the mean strain and rotation tensors, $S_{ij}$ and $\Omega_{ij}$.
- Calculate invariants, $II_S$ and $II_\Omega$
- Calculate $N$
- Calculate $\beta_1$ and $\beta_4$
- Calculate $a = \beta_1 T_1 + \beta_4 T_4$
- From $a_{ij}$ and $k$ we can obtain all the turbulence fluxes, which can be used to solve the mom'm equation.
- Finally, the auxillary $k$ and $\varepsilon$ equation are solved with the new momentum solution.

In the simplified 2D model, $N$ is obtained from the invariants:

<img src="N_2D_taulbee.png"/>

, where $c_1' = 1.8$. Next, the tensor coefficients are calculated as:

$$
\beta_1 = - \frac{6}{5} \frac{N}{N^2 - 2 II_\Omega} \quad , \quad \beta_4 = - \frac{6}{5} \frac{1}{N^2 - 2 II_\Omega} = \beta_1/N 
$$

Insert the coefficients in Pope's equation and isolate Reynolds stress:

$$
\begin{align}
\overline{u_i u_j} &= a_{ij} k + \frac{2}{3} k \delta_{ij} \\
        &= \left( \beta_1 T_1 + \beta_4 T_4 \right) k + \frac{2}{3} k \delta_{ij} \\
        &= - 2 \underbrace{\left(- \frac{1}{2} \beta_1\right)}_{C_{\mu,eff}} k T_1  + k \underbrace{\beta_4 T_4}_{a_{ij}^{(ex)}} + \frac{2}{3} k \delta_{ij}
\end{align}
$$



The model is equivalent to the usual Boussinesq hypothesis, but with an "effective" $C_\mu$ and an additional source term (this source should be treated explicitly to not cause troubles). This makes it relatively simple to implement!

## Self-conistent general 2D EARSM (Wallin 2000)

See appendix C of Wallin (2000). Has non-zero $\beta_1$, $\beta_2$ and $\beta_4$.


## Apsley and Leschziner EARSM family

Developed by Apsley and Leschziner (1998) and reviewed by van der Laan (2014). It is only applicable to flows, where the background flow is in equilibrium (i.e. log-law background flow). Some notation:
- They use "$G$" instead of "$\beta$"
- Apsley+Leschziner tensor group is used
- Their invaraints are called $\eta_1 , \dots , \eta_5$

The model comes in four versions (aka "orders"):

<img src="aps_lesch_coef.png"/>

Insert the coefficients in Pope's equation and isolate Reynolds stress:

$$
\begin{align}
\overline{u_i u_j} &= a_{ij} k + \frac{2}{3} k \delta_{ij} \\
        &= \left( G_1 T_1 + \dots + G_{10} T_{10} \right) k + \frac{2}{3} k \delta_{ij} \\
        &= \left( G_1 T_1 + \dots + II_\Omega G_6 S +  \dots + G_{10} T_{10} \right) k + \frac{2}{3} k \delta_{ij} \\
        &= - 2 \underbrace{\left(- \frac{1}{2} \beta_1\right)}_{C_{\mu,eff}} k T_1  + k \underbrace{\beta_4 T_4}_{a_{ij}^{(ex)}} + \frac{2}{3} k \delta_{ij}
\end{align}
$$



<img src="aps_lesch_cal.png"/>

All of the versions require the $f_P$-correction suggested by Apsley (1998):

$$
f_P = \frac{2 f_0}{1 + \sqrt{1 + 4 f_0 (f_0 -1) (\sigma/\tilde{\sigma})^2}} \quad , \quad f_0 = 1 + \frac{1 }{(C_1 - 1)}\quad , \quad \tilde{\sigma}=C_\mu^{-1/2} 
$$

Here $\tilde{\sigma}$ denotes the value of $\sigma$ evaluated in the log-law background flow, hence $\sigma/\tilde{\sigma} = 1$ and $f_P=1$ in the log-law region. 

$A_1$ and $A_2$ are calibration parameters, which can be set using a reference background flow (denoted by "tilde"):

$$
A_1 \equiv \frac{\tilde{a}_{11} + \tilde{a}_{33}}{\tilde{a}_{13}} \quad , \quad A_2 \equiv \frac{\tilde{a}_{11} - \tilde{a}_{33}}{\tilde{a}_{13}}
$$

ABL log-law flow (van der Laan (2014), p.28) with $C_\mu = 0.03$:
- $\tilde{a}_{11} = 0.38$, $\tilde{a}_{33} = -0.38$, $\tilde{a}_{13} = -\sqrt{C_\mu} = -0.173$ $\Rightarrow$ $A_1 = 0.0$, $A_2 = -4.388$ 

For comparison, the log-law region in a DNS of channel-flow is (c.f. Table 1 of Wallin (2000)): 
- $\tilde{a}_{11} = 0.34$, $\tilde{a}_{33} = -0.26$, $\tilde{a}_{13} = -0.29$ $\Rightarrow$ $A_1 = -0.276$, $A_2 = -2.069$ 

It can be noted that there is a "$-II_\Omega S$"-term in $T^{(6)}$, hence it contributes to $C_{mu,eff}$ defined from:

$$
a = - 2 C_{\mu,eff} S + ...
$$

Therefore $C_{\mu,eff} = - \frac{1}{2} \left( G^{(1)} - G^{(6)} II_\Omega\right)$. 


## Apsley and Leschziner 2D EARSM

I have not found an article that shows the 2D version of the model, but the 3D model can of course also be applied directly for 2D flows. However, if one wants to simplify the coding, two simplifications can be made in the code:

- $\eta_3 = \eta_4 = \eta_5 = 0$, which however only simplifies the $G^{(1)}$ coefficient of the 4'th order model..
- The following tensor groups are 0: $T_5$ and more?



## Apsley and Leschziner NLEVM models

(disclaimer: this is *NOT* the same as the Apsley-Leschziner EARSM)

Paul truncated the Apsley and Leschziner EARSM and only used the $T_1$ group, i.e.:

<img src="paul_truncation.png"/>

This can be re-written to:

$$
\begin{align}
\overline{u_i u_j} &= - 2 C_{\mu,eff} k T_1  + \frac{2}{3} k \delta_{ij}
\end{align}
$$

, where $C_{\mu,eff} = - \frac{1}{2} G^{(1)}$ is simply taken from *the first row* of the $G$-table from before. For all orders $C_{\mu,eff}$ is non-constant, because of $f_P$. One can note that order 1 and 2 are the same in this truncated mode, and actually these orders are what has become known as the $k$-$\varepsilon$-$f_P$ model.


## Self-consistent general 2D case

Here we *don't* assume $c_2=5/9$, so we can not throw away the third term of the ARSM equation $\rightarrow$ it gives much more complicated equations! 

Grimanji (1996a,b) was the first to find the analytical solution for $N$ and Wallin (2000) appendix C also gives it.


## Self-consistent simplified 3D case

Wallin (2000) shows the general 3D case, eq.2.10.



# Appendices

## App A: Derivation of the self-consistent simplified 2D model

This derivation follows the lines of Wallin (2000), but with more details. 

The "simplification" refers to:

- Use $c_2 = 5/9$ to throw away the last term of implicit equation (the "Taulbee simplification"). Divide through by $4/9$ and define $N = \left(c_1 - 1 + \frac{\mathcal{P}}{\varepsilon}\right) 9/4$ to give the "Taulbee equation":

$$
N a = - \frac{6}{5} S + \left( a \Omega - \Omega a \right)
$$

To start, we consider Pope's universal function, which in 2D is:
$$
a = \beta_1 T_1 + \beta_2 T_2 + \beta_4 T_4
$$

Then insert Pope's $a$ on RHS:

$$
\begin{align}
RHS &= - \frac{6}{5} S + \left( [\beta_1 T_1 + \beta_2 T_2 + \beta_4 T_4] \Omega - \Omega [\beta_1 T_1 + \beta_2 T_2 + \beta_4 T_4] \right) \\
    &= - \frac{6}{5} S + \beta_1 (T_1 \Omega - \Omega T_1) + \beta_2 (T_2 \Omega - \Omega T_2) + \beta_4 (T_4 \Omega - \Omega T_4) \\
    &= - \frac{6}{5} S + \beta_1 T_4 + \beta_2 \left([S^2 - \frac{1}{3} II_S \mathbf{I}] \Omega - \Omega [S^2 - \frac{1}{3} II_S \mathbf{I}]\right) + \beta_4 \left([S \Omega - \Omega S] \Omega - \Omega [S \Omega - \Omega S] \right) \\
    &= - \frac{6}{5} S + \beta_1 T_4 + \beta_2 (S^2 \Omega - \Omega S^2) + \beta_4 (S \Omega^2 - \Omega S \Omega - \Omega S \Omega + \Omega^2 S)
\end{align}
$$

Now we need some of Pope's 1975 appendix A expressions to proceed:

$$
\boxed{S^2 = \frac{1}{2} II_S \mathbf{I}  \quad \textrm{valid for 2D} }
$$

$$
\boxed{\Omega^2 = \frac{1}{2} II_\Omega \mathbf{I}  \quad \textrm{valid for 2D} }
$$

$$
\boxed{\Omega S \Omega = - \frac{1}{2} S  II_\Omega   \quad \textrm{valid for 2D} }
$$

Inserting these finally yields:

$$
\begin{align}
RHS &= - \frac{6}{5} S + \beta_1 T_4  + \beta_4 \left(2 S II_\Omega \right) \\
&= - \frac{6}{5} T_1 + \beta_1 T_4  + \beta_4 \left(2 T_1 II_\Omega \right) \\
&= \left(- \frac{6}{5} + 2 \beta_4 II_\Omega \right) T_1 + 0 \cdot T_2 + \beta_1 T_4  
\end{align}
$$

Compare with LHS:

$$
LHS = N \beta_1 T_1 + N \beta_2 T_2 + N \beta_4 T_4 ,
$$

to give the system of equations:

$$
N \beta_1 = \left(- \frac{6}{5} + 2 \beta_4 II_\Omega \right)
$$

$$
N \beta_2 = 0
$$

$$
N \beta_4 = \beta_1
$$


$$
\Rightarrow \beta_1 = - \frac{6}{5} \frac{N}{N^2 - 2 II_\Omega} \quad , \quad \beta_2 = 0 \quad , \quad \beta_4 = - \frac{6}{5} \frac{1}{N^2 - 2 II_\Omega} = \beta_1/N 
$$

We are done! Using the $\beta$-coefficients we have an explicit equation for $a$!! Only thing missing is an expression for $N$; this can be achieved by using $P/\varepsilon = - tr(aS)$. The analytic, closed expression is:

<img src="N_2D_taulbee.png"/>

Here $c_1' = 9/4 ( c_1 - 1)$ and $c_1 = 1.8$.

In summary the model is applied as:

- From last iteration, we have $U$, $V$, $k$ and $\varepsilon$ $\rightarrow$ we can get the mean strain and rotation tensors, $S_{ij}$ and $\Omega_{ij}$.
- Calculate invariants, $II_S$ and $II_\Omega$
- Calculate $N$
- Calculate $\beta_1$ and $\beta_4$
- Calculate $a = \beta_1 T_1 + \beta_4 T_4$
- Now we know the turbulence fluxes, which can be used to solve the mom'm equation.
- Finally, the auxillary $k$ and $\varepsilon$ equation are solved with the new momentum solution.

Can also pose the model as the normal Boussinesq hypothesis with an "effective" $C_\mu$ and an additional explicit term:

$$
\overline{u_i u_j} = - 2 C_{\mu,eff} k S_{ij} + \frac{2}{3} k \delta_{ij} + k a_{ij}^{(ex)}
$$

Here $S_{ij}$ is the dimensionless shear rate and 


$$
C_{\mu,eff} = - \frac{1}{2} \beta_1
$$

$$
a^{(ex)} = \beta_4 (S \Omega - \Omega S)
$$




## App B: Numerical implementation of Aps+Lesch 3D tensor groups

The 10 tensor groups of Aps+Lesch were shown in the start of the document. The strain and rotation tensors:

$$
S_{ij} = \begin{bmatrix} S_{11} & S_{12} & S_{13} \\ S_{12} & S_{22} & S_{23} \\ S_{13} & S_{23} & S_{33} \end{bmatrix}  \quad , \quad
\Omega_{ij} = \begin{bmatrix} 0 & \Omega_{12} & \Omega_{13} \\ -\Omega_{12} & 0 & \Omega_{23} \\ -\Omega_{13} & -\Omega_{23} & 0 \end{bmatrix} 
$$


In [1]:
import sympy as sym
# Method to display results like "idx = symObj"
#https://stackoverflow.com/questions/46993589/modify-jupyter-notebooks-ipython-kernel-latex-rendering-of-sympy-objects
from sympy import latex
from IPython.display import display_latex
def disp(idx, symObj):
    eqn = '\\[' + idx + ' = ' + latex(symObj) + '\\]'
    display_latex(eqn,raw=True)
    return

# All tensor groups are symmetric, so we only need upper part of matrices
s11, s22, s33, s12, s13, s23, o12, o13, o23 = sym.symbols('s11, s22, s33, s12, s13, s23, o12, o13, o23')

# Normalized strain tensor (dimensionless)
s = sym.Matrix([[s11, s12, s13],[s12, s22, s23],[s13, s23, s33]])

# Normalized rotation tensor  (dimensionless)
o = sym.Matrix([[0, o12, o13],[-o12, 0, o23],[-o13, -o23, 0]])

# Define symbolic s^2 to help simplify later expressions
ss11, ss22, ss33, ss12, ss13, ss23 = sym.symbols('ss11, ss22, ss33, ss12, ss13, ss23')
ss = sym.Matrix([[ss11, ss12, ss13],[ss12, ss22, ss23],[ss13, ss23, ss33]])

# Define symbolic o^2 to help simplify later expressions
oo11, oo22, oo33, oo12, oo13, oo23 = sym.symbols('oo11, oo22, oo33, oo12, oo13, oo23')
oo = sym.Matrix([[oo11, oo12, oo13],[oo12, oo22, oo23],[oo13, oo23, oo33]])

# Define symbolic os to help simplify later expressions. Note O.S is NOT symmetric
os11, os12, os13, os21, os22, os23, os31, os32, os33 = sym.symbols('os11, os12, os13, os21, os22, os23, os31, os32, os33')
os = sym.Matrix([[os11, os12, os13],[os21, os22, os23],[os31, os32, os33]])

# T2 = os-so:
T2 = o*s-s*o

# T3 = s^2-1/3tr(s^2)I:
T3 = s*s-sym.Rational(1,3)*sym.trace(s*s)*sym.eye(3)

# T4 = o^2-1/3tr(o^2)I:
T4 = o*o-sym.Rational(1,3)*sym.trace(o*o)*sym.eye(3)

# T5 = o*s^2 - s^2*o 
T5 = o*ss - ss*o

# T6-2/3*IV*I-tr(oo)*s = s*o^2 + o^2*s
T6min = s*oo + oo*s

disp('S_{ij}',s)
disp('\Omega_{ij}',o)
disp('S^2',s*s)
disp('S^2',ss)
disp('\Omega^2',o*o)
disp('\Omega^2',oo)
disp('\Omega S',o*s)
disp('\Omega S',os)
disp('\Omega S^2',os*s)
disp('T_2 = \Omega S - \Omega S',T2)
disp(r'T_3 = S^2 - \frac{1}{3} II_S I',T3)
disp(r'T_4 = \Omega^2 - \frac{1}{3} II_\Omega I',T4)
disp(r'T_5 = \Omega S^2 - S^2 \Omega',T5)
disp(r'T_6 - II_\Omega S - 2/3 IV I = S \Omega^2 - \Omega^2 S',T6min)

In [2]:
# numpy test of T6min for parallel flow
sigma = 2*2.026
import numpy as np
A = np.array([[0, 0, 0.5*sigma],[0, 0, 0], [0.5*sigma, 0, 0]])
B = np.array([[0, 0, 0.5*sigma],[0, 0, 0], [-0.5*sigma, 0, 0]])
np.matmul(A,np.matmul(B,B)) - np.matmul(B,np.matmul(B,A))

array([[0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.]])

**The last groups are quite long, so use tensor mathematics to proof relations**

#### Group $T_7$

$$
T^{(7)} = \Omega S \Omega^2 - \Omega^2 S \Omega
$$

The second term can be re-phased:

$$
\Omega^2 S \Omega = \Omega_{ik} \Omega_{km} S_{mn} \Omega_{nj} = - \Omega_{jn} S_{nm} \Omega_{mk}  \Omega_{ki} = - (\Omega S \Omega^2)^T 
$$

Hence,

$$
T^{(7)}_{ij} = (\Omega S \Omega^2)_{ij} + (\Omega S \Omega^2)_{ji} 
$$

The transpose does nothing to the diagonal, so the diagonal terms are simply: 

$$T^{(7)}_{\alpha\alpha} = 2(\Omega S \Omega^2)_{\alpha\alpha} = 2(\Omega S)_{\alpha i} (\Omega^2)_{i \alpha} = 2(\Omega S)_{\alpha i} (\Omega^2)_{\alpha i}$$

The off-diagonal terms can be calculated using the two matrices, $\Omega S$ and $\Omega^2$: 

$$T^{(7)}_{\alpha\beta} = (\Omega S \Omega^2)_{\alpha\beta} + (\Omega S \Omega^2)_{\beta\alpha}  = (\Omega S)_{\alpha i} (\Omega^2)_{i \beta } + (\Omega S)_{\beta i} (\Omega^2)_{i \alpha } = (\Omega S)_{\alpha i} (\Omega^2)_{\beta i } + (\Omega S)_{\beta i} (\Omega^2)_{\alpha i}  $$

#### Group $T_8$


$$
T^{(8)} = S \Omega  S^2 - S^2 \Omega S 
$$

The first term can be re-phased:

$$
S \Omega  S^2 = - (S^2 \Omega S)^T 
$$

Hence,

$$
T^{(8)}_{ij} = - ( S^2 \Omega S)_{ji} - (S^2 \Omega S)_{ij} 
$$

The transpose does nothing to the diagonal, so the diagonal terms are simply: 

$$T^{(8)}_{\alpha\alpha} = - 2( S^2 \Omega S)_{\alpha\alpha} = - 2(S^2)_{\alpha i} (\Omega S)_{i \alpha} $$

The off-diagonal terms can be calculated using the two matrices, $\Omega S$ and $S^2$: 

$$T^{(8)}_{\alpha\beta} =  - (S^2 \Omega S)_{\beta\alpha} -(S^2 \Omega S)_{\alpha\beta} =  - (S^2)_{\beta i} (\Omega S)_{i \alpha }  - (S^2)_{\alpha i} (\Omega S)_{i \beta }  $$


#### Group $T_9$


$$
T^{(9)} = S^2 \Omega^2 + \Omega^2 S^2 - \frac{2}{3} V \mathbf{I}
$$

The first term can be re-phased:

$$
S^2 \Omega^2 = (\Omega^2 S^2 )^T 
$$

Hence,

$$
T^{(9)}_{ij} + \frac{2}{3} V \mathbf{I} = ( \Omega^2 S^2)_{ji} + (\Omega^2 S^2)_{ij} 
$$

The transpose does nothing to the diagonal, so the diagonal terms are simply: 

$$T^{(9)}_{\alpha\alpha} = 2( \Omega^2 S^2 )_{\alpha\alpha} =  2(\Omega^2)_{\alpha i} (S^2)_{i \alpha} $$

The off-diagonal terms can be calculated using the two matrices, $\Omega^2$ and $S^2$: 

$$T^{(9)}_{\alpha\beta} =   (\Omega^2 S^2)_{\beta\alpha}  + (\Omega^2 S^2)_{\alpha\beta} =   (\Omega^2)_{\beta i} (S^2)_{i \alpha }  + (\Omega^2)_{\alpha i} (S^2)_{i \beta }  $$


#### Group $T_{10}$

$$
T^{(10)} = \Omega S^2 \Omega^2 - \Omega^2 S^2 \Omega
$$

The second term can be re-phased:

$$
\Omega^2 S^2 \Omega = - (\Omega S^2 \Omega^2)^T 
$$

Hence,

$$
T^{(10)}_{ij} = (\Omega S^2 \Omega^2)_{ij} + (\Omega S^2 \Omega^2)_{ji} 
$$

The transpose does nothing to the diagonal, so the diagonal terms are simply: 

$$
\begin{align}
T^{(10)}_{\alpha\alpha} &= 2(\Omega S^2 \Omega^2)_{\alpha\alpha} \\
                        &= 2(\Omega S^2)_{\alpha i} (\Omega^2)_{i \alpha} \\
                        &= 2(\Omega S^2)_{\alpha i} (\Omega^2)_{\alpha i}
\end{align}
$$

The off-diagonal terms can be calculated using the two matrices, $\Omega S^2$ and $\Omega^2$: 

$$
\begin{align}
T^{(10)}_{\alpha\beta} &= (\Omega S \Omega^2)_{\alpha\beta} + (\Omega S \Omega^2)_{\beta\alpha} \\
                       &= (\Omega S^2)_{\alpha i} (\Omega^2)_{i \beta } + (\Omega S^2)_{\beta i} (\Omega^2)_{i \alpha }\\
                       &= (\Omega S^2)_{\alpha i} (\Omega^2)_{\beta i } + (\Omega S^2)_{\beta i} (\Omega^2)_{\alpha i}  
\end{align}
$$

In [3]:
# Here is a numerical proof of the T7 group rule:
# T7= o*s*o^2 - o^2*s*o
A = sym.Matrix([[4, 2, 1.23],[2, 3.3, 5.7],[1.23, 5.7, -7.3]])
B = sym.Matrix([[0, 2.5, 1.66],[-2.5, 0, 4.7],[-1.66, -4.7, 0]])
disp('A^2 B A + (A B A^2)^T',A*A*B*A  + (A*B*A*A).T)
# This shows that A*A*B*A = -(A*B*A*A).T


## Invariant 3 = tr(S.S.S), which can be calculated from S^3.
disp('S^3',s*ss)

# The helper variable for calculating T10
disp('\Omega S^2',os*s)

# App c: Parallel flow

In [4]:
sig = sym.symbols('\sigma')
S = sym.Matrix([[0, 0, sym.Rational(1,2)*sig],[0, 0, 0],[sym.Rational(1,2)*sig, 0, 0]])
O = sym.Matrix([[0, 0, sym.Rational(1,2)*sig],[0, 0, 0],[-sym.Rational(1,2)*sig, 0, 0]])
disp('S_{ijPar}',S)
disp('\Omega_{ijPar}',O)

### Invariants

In [12]:
def sym_invariants(S,O):
    eta1 = sym.trace(S*S)
    eta2 = sym.trace(O*O)
    eta3 = sym.trace(S*S*S)
    eta4 = sym.trace(S*O*O)
    eta5 = sym.trace(S*S*O*O)
    disp('II_S',eta1)
    disp('II_\Omega',eta2)
    disp('III',eta3)
    disp('IV',eta4)
    disp('V',eta5)
    
sym_invariants(S,O)

### Tensors (Aps+Lesch groups)

In [21]:
def sym_tensors(S,O):
    '''
    Calculate tensors of Apsley+Leschziner symbolically
    '''
    eta2 = sym.trace(O*O)
    eta4 = sym.trace(S*O*O)
    eta5 = sym.trace(S*S*O*O)
    T1 = S
    T2 = O*S-S*O
    T3 = S**2 - sym.Rational(1,3)*sym.trace(S**2)*sym.eye(3)
    T4 = O**2 - sym.Rational(1,3)*sym.trace(O**2)*sym.eye(3)
    T5 = O*S*S - S*S*O
    T6 = S*O*O + O*O*S - eta2*S - sym.Rational(2,3)*eta4*sym.eye(3)
    T7 = O*S*O*O - O*O*S*O
    T8 = S*O*S*S - S*S*O*S
    T9 = O*O*S*S + S*S*O*O - sym.Rational(2,3)*eta5*sym.eye(3)
    T10 = O*S*S*O*O - O*O*S*S*O
    # Use simplify, because standard sympy doesn't cancel some terms..
    disp('T_1',T1)
    disp('T_2',T2)
    disp('T_3',T3)
    disp('T_4',T4)
    disp('T_5',sym.simplify(T5))
    disp('T_6',sym.simplify(T6))
    disp('T_7',sym.simplify(T7))
    disp('T_8',sym.simplify(T8))
    disp('T_9',sym.simplify(T9))
    disp('T_{10}',sym.simplify(T10))
    
sym_tensors(S,O)

It can be noted that:

$$
\begin{align}
T_4 &= -T_3 \\
T_7 &= - \frac{\sigma^2}{4} T_2 = \frac{1}{2} II_\Omega T_2  \\
T_8 &= - \frac{\sigma^2}{4} T_2 = \frac{1}{2} II_\Omega T_2\\
T_9 &= - \frac{\sigma^2}{2} T_3 = II_\Omega T_3  \\
\end{align}
$$

Hence, all groups can be written as a function of $T_1$, $T_2$ and $T_3$.

Also, only $T_1$ has non-zero non-diagonal elements. This means that $a_{13}^{ex} = 0$. This fact is important in 1D, because only $\frac{\partial a_{13}^{(ex)}}{\partial z}$ appears in the U-mom'm eq. Therefore there is no anisotropy mom'm source in 1D! I choose to keep it in the code for generality.

### Tensors (Wallin groups)

The sixth group does NOT have $II_\Omega S$:

In [7]:
T6wal = S*O*O + O*O*S - sym.Rational(2,3)*eta4*sym.eye(3)
disp('T_{wal6}',T6wal)

This is equivalent to:

$$
T_{wal6} = - \frac{\sigma^2}{2} T_1 = II_\Omega T_1 
$$

## App D: EARSM 1D (with veering)

Assume the only non-zero gradients are: 

$$
\frac{\partial U}{\partial z}, \frac{\partial V}{\partial z}
$$ 

This is actually NOT a subcase of 2D flows...

$$
S_{ij} = \begin{bmatrix} 0 & 0 & \frac{1}{2} \frac{\partial U}{\partial z}  \\ 0 & 0 & \frac{1}{2} \frac{\partial V}{\partial z} \\ \frac{1}{2} \frac{\partial U}{\partial z}  & \frac{1}{2} \frac{\partial V}{\partial z} & 0 \end{bmatrix}  \quad , \quad
\Omega_{ij} = \begin{bmatrix} 0 & 0 & \frac{1}{2} \frac{\partial U}{\partial z}  \\ 0 & 0 & \frac{1}{2} \frac{\partial V}{\partial z} \\ -\frac{1}{2} \frac{\partial U}{\partial z}  & -\frac{1}{2} \frac{\partial V}{\partial z} & 0 \end{bmatrix}  
$$

In [22]:
sigu = sym.symbols(r'\sigma_u')
sigv = sym.symbols(r'\sigma_v')
Sv = sym.Matrix([[0, 0, sigu],[0, 0, sigv],[sigu, sigv, 0]])
Ov = sym.Matrix([[0, 0, sigu],[0, 0, sigv],[-sigu, -sigv, 0]])
disp('S_{ijVeer}',Sv)
disp('\Omega_{ijVeer}',Ov)
disp('S_{ijVeer}^2',Sv**2)
disp('\Omega_{ijVeer}^2',Ov**2)
sym_invariants(Sv,Ov)
sym_tensors(Sv,Ov)

There is some similarities with the non-veered case:
- $T_5$, $T_6$ $T_{10}$ are all 0's.
- Only $T_1$ has non-zero 13 and 23 components. The u- and v-eqs still only depend on $a_{13}^{(ex)}$ and $a_{23}^{(ex)}$, therefore we can again say that there is no extra anisotropy mom'm source term in the mom'm eqs!

## App E: EARSM 2D

Assume no variation in y-direction and $V=0$ (Pope 1975, p.334); this is the definition of 2D. The non-zero gradients are: 

$$
\frac{\partial U}{\partial x}, \frac{\partial U}{\partial z}, \frac{\partial W}{\partial x}, \frac{\partial W}{\partial z} 
$$ 

But pr. continuity $\frac{\partial W}{\partial z} = - \frac{\partial U}{\partial x}$.

$$
S_{ij} = \begin{bmatrix} \frac{\partial U}{\partial x} & 0 & \frac{1}{2} \left(\frac{\partial U}{\partial z} + \frac{\partial W}{\partial x} \right) \\ 0 & 0 & 0 \\ \frac{1}{2} \left(\frac{\partial U}{\partial z} + \frac{\partial W}{\partial x} \right) & 0 & -\frac{\partial U}{\partial x} \end{bmatrix}  \quad , \quad
\Omega_{ij} = \begin{bmatrix} 0 & 0 & \frac{1}{2} \left(\frac{\partial U}{\partial z} - \frac{\partial W}{\partial x} \right) \\ 0 & 0 & 0 \\ -\frac{1}{2} \left(\frac{\partial U}{\partial z} - \frac{\partial W}{\partial x} \right) & 0 & 0 \end{bmatrix}  
$$

In [8]:
Ux = sym.symbols(r'\frac{\partial~U}{\partial~x}')
Uz = sym.symbols(r'\frac{\partial~U}{\partial~z}')
Wx = sym.symbols(r'\frac{\partial~W}{\partial~x}')
S2d = sym.Matrix([[Ux, 0, 0.5*(Uz+Wx)],[0, 0, 0],[0.5*(Uz+Wx), 0, -Ux]])
O2d = sym.Matrix([[0, 0, 0.5*(Uz-Wx)],[0, 0, 0],[0.5*(Wx-Uz), 0, 0]])
eta1 = sym.trace(S2d**2)
eta2 = sym.trace(O2d**2)
eta3 = sym.trace(S2d**3)
eta4 = sym.trace(O2d*S2d*S2d)
eta5 = sym.trace(O2d*O2d*S2d*S2d)
disp('S_{ij2D}',S2d)
disp('\Omega_{ij2D}',O2d)
disp('S_{ij2D}^2',S2d**2)
disp('\Omega_{ij2D}^2',O2d**2)
disp('II_S',eta1)
disp('II_\Omega',eta2)
disp('III',eta3)
disp('IV',eta4)
disp('V',eta5)

In [9]:
e1 = eta1.evalf(subs={Uz: -0.3, Wx: 0.1, Ux: 0.2, })
e2 = eta2.evalf(subs={Uz: 0.3, Wx: -0.1})
print('e1 = %.4f and e2 = %.4f'%(e1,e2) )


e1 = 0.1000 and e2 = -0.0800


Conclusions:

- $II_S \neq - II_\Omega$. This is in contrast to parallel flow.
- It seems like $II_S > 0$ always.
- It seems like $II_\Omega <0$ always

## Turbulence $k$-$\varepsilon$ modelling constants

Even EARSM needs modelling constants. $\sigma_k = 1.0$ and $\sigma_\varepsilon = 1.3$ are used in almost all litterature, but there is some freedom among the others. They should however satisfy, c.f. (Richards 1993, eq.10), Sørensen (1995, eq.10) or (van der Laan 2017, eq.23):

$$
C_{\varepsilon,1} + \kappa^2 C_\mu^{-1/2} \sigma_{\varepsilon}^{-1} = C_{\varepsilon,2}
$$




In [11]:
sigma_eps = 1.37
kappa = 0.4
cmu = 0.06
ceps1 = 1.44

ceps2 = ceps1 + kappa**2*cmu**(-0.5)*sigma_eps**(-1)

print('ceps2 = %.5f'%(ceps2))

ceps2 = 1.91679


# [Other notebook](./testA.ipynb)

<!--
|                            	| $\kappa$ 	| $C_\mu$ 	| $\sigma_k$ 	| $\sigma_\varepsilon$ 	| $C_{\varepsilon,1}$ 	| $C_{\varepsilon,2}$ 	|
|----------------------------	|----------	|---------	|------------	|----------------------	|---------------------	|---------------------	|
| Sørensen (1995) ABL        	| 0.4      	| 0.03    	| 1.0        	| 1.3                  	| 1.21                	| 1.92                	|
| Sørensen (1995) Industrial 	| 0.4      	| 0.09    	| 1.0        	| 1.3                  	| 1.42                	| 1.92                	|
| Velibor (2019) BC article  	| 0.4      	| ?       	| 1.0        	| 1.3                  	| 1.44                	| 1.92                	|
| Velibor (2021) thesis      	| 0.4      	| 0.09    	| 1.0        	| 1.3                  	| 1.44                	| 1.82                	|

asf
-->